Integrate cuBLASLt API into backend

benbarsdell opened the pull request:
The cuBLASLt API provides more features and better control over the execution of GEMMs than the regular cuBLAS API.

This PR adds the following:

- Build integration and stream executor wrappers for the cublasLt API.
- Integration (with autotuning) into the batch_matmul_op_impl class (used by the BatchMatMul and Einsum ops).
- Support for bias + ReLU fusion, which is expected to be useful for XLA in the future.

(I note that this is a large PR and it may be easier to review the commits separately. Also the diff is a bit strange in places due to some minor refactoring I had to do to simplify the code).

cc @reedwm @nluehr

benbarsdell committed:
Add cublasLt wrappers to stream_executor

- Adds ThenBlasLtMatmul routines that behave similarly to
  ThenBlasGemmWithAlgorithm but call into the cublasLt library and allow
  separation of plan creation and execution.
- A list of heuristically-prioritized opaque algorithm objects can be
  obtained via GetBlasLtMatmulAlgorithms.
- These routines are only supported when the CUDA version is >= 11.0.

benbarsdell committed:
Use BlasLtMatmul APIs in batch_matmul_op_impl

- Integrates BlasLtMatmul with autotuning into the implementation of the
  BatchMatMul and Einsum ops.
- This integration is only used when the CUDA version is >= 11.0.

benbarsdell committed:
Remove duplicated code in conv and fft kernels

benbarsdell committed:
Add workaround for cublasLt known issue

- Avoids a heuristic alignment issue noted in the CUDA Release Notes.

benbarsdell committed:
Add support for blasLt epilogue fn and bias vector

- Changes the backend APIs to allow an epilogue function (default, ReLU,
  bias, or bias then ReLU) to be specified and a bias vector to be
  provided.
- This is expected to be useful for XLA to perform fusions.
- This functionality is not currently tested, because the BatchMatMulOp
  does not expose relu/bias fusion.

benbarsdell commented:
I am investigating the build issues.

benbarsdell commented:
I think I've fixed all the build issues now.

timshen91 commented on tensorflow/stream_executor/blas.h:
```diff
@@ -101,13 +101,72 @@ enum class ComputationType {
   kComplexF64,  // Complex number comprised of two f64s.
+  // The below values are only supported for BlasLt routines (both real and
+  // complex).
+  kF32FastTF32, // 32-bit floating-point with reduced (>=10-bit) mantissa
```
Can they be "kTF32" and "kBF16"?

benbarsdell replied:
The naming is a bit tricky here. Both use F32 accumulation, with the difference being the mantissa size that the input is rounded to (I've added a comment describing this behavior in more detail). For this reason I'm not sure that TF32/BF16 would be sufficient, though I'm open to changing them if you think it would be better.

benbarsdell replied:
Done

timshen91 commented on tensorflow/stream_executor/blas.h:
```diff
@@ -101,13 +101,72 @@ enum class ComputationType {
 std::ostream &operator<<(std::ostream &os, ComputationType ty);
 
+// Type with which inputs and outputs of a blaslt routine are performed.
+enum class DataType {
```
dnn.proto also has DataType. Maybe pull it out of the dnn namespace and reuse it?

benbarsdell replied:
Done (I just did it by including dnn.h; let me know if this is undesirable).

timshen91 commented on tensorflow/stream_executor/blas.h:
```diff
@@ -1383,6 +1455,93 @@ class BlasSupport {
+  // can then be passed to DoBlasLtMatmul(). When possible, plans should be
+  // created once and reused for multiple calls to DoBlasLtMatmul().
+  // Returns a null pointer on failure.
+  std::unique_ptr<blas::IBlasLtMatmulPlan> CreateBlasLtMatmulPlan(
```
Can we create a `struct MatmulPlanParams` for it? Something like:
```
struct MatmulPlanParams {
  blas::DataType ab_type;
  blas::DataType c_type;
  ...
  int64 lda;
  int64 stride_a = 0;
  int64 ldb;
  int64 stride_b = 0;
  int64 ldc;
  int64 stride_c = 0;
};
```

This way the function signature is simpler and can be updated all at once through extending MatmulPlanParams.

benbarsdell replied:
Done

timshen91 commented on tensorflow/stream_executor/blas.h:
```diff
@@ -1383,6 +1455,93 @@ class BlasSupport {
+      uint64 k, int64 lda, int64 ldb, int64 ldc) {
+    return CreateBlasLtMatmulPlanStridedBatched(
+        ab_type, c_type, computation_type, pointer_mode, transa, transb, m, n,
+        k, 1, lda, 0, ldb, 0, ldc, 0);
```
Would the overloading still be necessary if we have `MatmulPlanParams` above with defaulted strides?

benbarsdell replied:
Removed

timshen91 commented on tensorflow/stream_executor/blas.h:
```diff
@@ -1383,6 +1455,93 @@ class BlasSupport {
+  // is not nullptr, the operation is profiled, error messages are
+  // suppressed, and output_profile_result->algorithm() is set to
+  // algorithm->index().
+  virtual bool DoBlasLtMatmul(
```
Are we able to remove all these overloadings by similarly using a struct for parameters?

benbarsdell replied:
Removed by using runtime types instead of static type overloads.

timshen91 commented on tensorflow/stream_executor/cuda/cublasLt_stub.cc:
```diff
@@ -0,0 +1,59 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
@chsigg Is this file supposed to be auto-generated?

timshen91 commented on tensorflow/stream_executor/stream.h:
```diff
@@ -1665,6 +1665,56 @@ class Stream {
                        const DeviceMemory<std::complex<double>> &a, int lda,
                        DeviceMemory<std::complex<double>> *b, int ldb);
 
+  // See BlasSupport::DoBlatLtMatmul.
```
ditto - can we simplify these overloadings using the parameter struct I mentioned above? Also, is it easy to make data types / compute types template parameters, or even runtime types? Again, I'd weigh towards simplicity of the signature, as these functions are called for maybe twice throughout Tensorflow.

benbarsdell replied:
I've now done this with runtime types the same way as in blas.h. I had to add some helper code to work around minor issues that came up with logging and type deduction.

benbarsdell replied:
I since realized that this wasn't actually calling the template version of DoBlasLtMatmul any more. I've now refactored it again so that it's a template implementation with explicit instantiations.

timshen91 commented on tensorflow/stream_executor/blas.h:
```diff
@@ -1383,6 +1455,93 @@ class BlasSupport {
+  // suppressed, and output_profile_result->algorithm() is set to
+  // algorithm->index().
+  virtual bool DoBlasLtMatmul(
+      Stream* stream, const blas::IBlasLtMatmulPlan* plan,
```
Can we not have compile-time data types overloaded? These are not public functions, and they will be called likely once. You can refer to dnn::DoConvolve as an example.

benbarsdell replied:
I've replaced these overloads with runtime types similar to DoConvolve. I had to add a dynamically-typed version of HostOrDeviceScalar to enable this.

gbaned commented:
@benbarsdell  Can you please check @timshen91's comments and keep us posted ? Thanks!

benbarsdell committed:
Replace CreateBlasLtMatmulPlan args with struct

benbarsdell committed:
Use (+generalize) existing dnn::DataType in blas::

benbarsdell committed:
Replace blasLt overloads with runtime types

benbarsdell committed:
Add more detailed comment for kF32FastTF32/BF16

benbarsdell committed:
Replace stream BlasLt overloads with runtime types

timshen91 commented:
I finished reviewing stream_executor/*, and will take a look later on tensorflow/core.

benbarsdell committed:
Fix formatting issue in cuda_blas.cc

benbarsdell committed:
Rename kF32FastTF32/BF32 to kTF32/BF16AsF32

benbarsdell committed:
Add type checks in DoBlasLtMatmul

benbarsdell committed:
Change ThenBlasLtMatmul implem to a template

- This is required to ensure that the template version of
  DoBlasLtMatmul is called (which is important because it performs
  additional type checks).

benbarsdell committed:
Refactor blasLt APIs to return Status, not bool

- This does not include DoBlasLtMatmul because the helpers in
  stream.cc require it to return bool.

benbarsdell committed:
Remove unneeded code for HostOrDeviceScalar<void>

benbarsdell commented on tensorflow/stream_executor/blas.h:
```diff
@@ -1383,6 +1447,58 @@ class BlasSupport {
+                      const blas::IBlasLtMatmulAlgorithm* algorithm,
+                      const DeviceMemory<CType>& bias = {},
+                      blas::ProfileResult* output_profile_result = nullptr) {
+    return DoBlasLtMatmul(stream, plan, alpha, a, b, beta, *c,
```
Done

benbarsdell commented on tensorflow/stream_executor/cuda/cuda_blas.cc:
```diff
@@ -2912,6 +3049,541 @@ bool CUDABlas::DoBlasTrsm(Stream *stream, blas::Side side,
+template <>
+blas::ComputationType ToComputationType<std::complex<float>>() {
+  return blas::ComputationType::kComplexF32;
+}template <>
```
Done

benbarsdell commented on tensorflow/stream_executor/cuda/cuda_blas.cc:
```diff
@@ -2912,6 +3049,541 @@ bool CUDABlas::DoBlasTrsm(Stream *stream, blas::Side side,
+namespace {
+
+template <typename T>
+inline bool SetCublasLtAttr(cublasLtMatrixLayout_t handle,
```
I've done this for `CreateBlasLtMatmulPlan` and `GetBlasLtMatmulAlgorithms` and the internal functions behind them.
Doing it for DoBlasLtMatmul is not so easy, as the helpers in stream.cc assume that all the DoBlas* functions return bool, so I've left that as-is.

benbarsdell committed:
Fix cuda_blas.cc compilation for old CUDA versions

timshen91 reviewed (approved):
Looks good now. About landing the PR: if it appears to be difficult to land (I need to run our internal benchmarks), we might want to land StreamExecutor first, and then TF. But we will see.

Thanks for the work!

timshen91 commented:
Hi @benbarsdell , I found a correctness regression on the following test: `//tensorflow/python/kernel_tests/distributions:dirichlet_test_gpu`

the error message looks like this:
```
======================================================================
FAIL: testCovarianceFromSampling (__main__.DirichletTest)
testCovarianceFromSampling (__main__.DirichletTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/distributions/dirichlet_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 1196, in decorated
    f(self, *args, **kwargs)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/distributions/dirichlet_test_gpu.runfiles/org_tensorflow/tensorflow/python/kernel_tests/distributions/dirichlet_test.py", line 201, in testCovarianceFromSampling
    self.assertAllClose(sample_stddev_, analytic_stddev, atol=0.02, rtol=0.)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/distributions/dirichlet_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 1236, in decorated
    return f(*args, **kwds)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/distributions/dirichlet_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 2698, in assertAllClose
    self._assertAllCloseRecursive(a, b, rtol=rtol, atol=atol, msg=msg)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/distributions/dirichlet_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 2658, in _assertAllCloseRecursive
    (path_str, path_str, msg)))
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/distributions/dirichlet_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 2593, in _assertArrayLikeAllClose
    a, b, rtol=rtol, atol=atol, err_msg="\n".join(msgs), equal_nan=True)
  File "/usr/local/lib/python3.6/site-packages/numpy/testing/_private/utils.py", line 1528, in assert_allclose
    verbose=verbose, header=header, equal_nan=equal_nan)
  File "/usr/local/lib/python3.6/site-packages/numpy/testing/_private/utils.py", line 840, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=0, atol=0.02
Mismatched value: a is different from b. 
not close where = (array([0, 0, 0, 1, 1]), array([0, 1, 2, 0, 1]))
not close lhs = [0.05104237 0.06437208 0.06834181 0.06432743 0.0643561 ]
not close rhs = [0.14085905 0.17817417 0.18898225 0.17747666 0.17760938]
not close dif = [0.08981667 0.11380209 0.12064043 0.11314923 0.11325329]
not close tol = [0.02 0.02 0.02 0.02 0.02]
dtype = float32, shape = (2, 3)
Mismatched elements: 5 / 6 (83.3%)
Max absolute difference: 0.12064043
Max relative difference: 0.6587895
 x: array([[0.051042, 0.064372, 0.068342],
       [0.064327, 0.064356, 0.004876]], dtype=float32)
 y: array([[0.140859, 0.178174, 0.188982],
       [0.177477, 0.177609, 0.014291]], dtype=float32)

```

There are other failures that look like precision differences, which only needs to increase the test tolerance. But for this failure, the values are `0.05104237` vs `0.14085905`, which is too large to be a precision issue.

This was run on a P100. I can not reproduce with a Volta GPU.

benbarsdell commented:
I have reproduced what looks like the same issue, on a Volta card. I'm not sure why the CI didn't fail.

It appears to be due to an issue in cublasLt and I am working on a workaround.

timshen91 commented:
Hi @benbarsdell , can you rebase this PR? I'm trying to run it with our internal benchmarks.
